services/search/pkg/search/service.go

```diff
@@ -487,6 +490,63 @@ func (s *Service) UpsertItem(ref *provider.Reference, uID *user.UserId) {
 	} else {
 		logDocCount(s.engine, s.logger)
 	}
+
+	// determine if metadata needs to be stored in storage as well
+	metadata := map[string]string{}
+	addAudioMetadata(metadata, doc.Audio)
+	if len(metadata) == 0 {
+		return
+	}
+
+	s.logger.Trace().Str("name", doc.Name).Interface("metadata", metadata).Msg("Storing metadata")
+
+	gatewayClient, err := s.gatewaySelector.Next()
+	if err != nil {
+		s.logger.Error().Err(err).Msg("could not retrieve client to store metadata")
+		return
+	}
+
+	resp, err := gatewayClient.SetArbitraryMetadata(ctx, &provider.SetArbitraryMetadataRequest{
+		Ref: ref,
+		ArbitraryMetadata: &provider.ArbitraryMetadata{
+			Metadata: metadata,
+		},
+	})
+	if err != nil || resp.GetStatus().GetCode() != rpc.Code_CODE_OK {
+		s.logger.Error().Err(err).Msg("error storing metadata")
+		return
+	}
+}
+
+func addAudioMetadata(metadata map[string]string, audio *libregraph.Audio) {
+	if audio == nil {
+		return
+	}
+	marshalToStringMap(audio, metadata, "libre.graph.audio.")
+}
+
+func marshalToStringMap[T libregraph.MappedNullable](source T, target map[string]string, prefix string) {
+	// ToMap never returns a non-nil error ...
+	m, _ := source.ToMap()
+
+	for k, v := range m {
+		target[prefix+k] = valueToString(v)
+	}
+}
+
+func valueToString(value interface{}) string {
```

Comment:
This will panic when I do:
```
var v *string
valueToString(v)
```

Reply:
I guess you simply shouldn't 😁 

... but yeah, I can probably add a check for nil values

Reply:
Still panicing unfortunately. See go playground example: https://play.golang.com/p/Szh074cVddu

We need to check for `v != nil` inside the type switch:
```
case *string:
		if v != nil {
			return *v
		}
```

Reply:
yeah, technically it's superfluous - ToMap() returns only non-nil values. I still added checks in the spirit of defensive programming